Add: Include history of cargo delivered and waiting to industry graphs.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Industry cargo history only shows cargo production and cargo transported away. These are probably the most useful statistics, but recording accepted and waiting (stock piled) cargo might also be interesting.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

This PR adds history to cargo accepted by industries.

* Total amount accepted is a simple counter, summing each delivery.
* Average waiting amount (stock pile) is instead a gauge. ~~A rolling average of the past 30 economy days is used~~ An average of the past month is accumulated and used ~~(even in non-wallclock mode)~~ to provide a single monthly value. This is sampled once every economy day, so highs and lows can be missed, but sampling every tick seems excessive.

![image](https://github.com/user-attachments/assets/33355a2f-c798-4493-852b-b04d21a339bf)

Bumps the savegame version.

~~This builds on top of #14309 and #14313. The graphs now contain more data. While each series can be hidden, the highlight feature becomes more useful.~~

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

Vanilla industries don't stock pile, so this part is only of use with add-on industries, and even then it might not be a feature of the set. I don't think there is a simple way to test if stock piles are used, as it's callback driven.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

2TallTyler commented on src/saveload/saveload.h:
```diff
@@ -403,6 +403,8 @@ enum SaveLoadVersion : uint16_t {
 	SLV_FIX_SCC_ENCODED_NEGATIVE,           ///< 353  PR#14049 Fix encoding of negative parameters.
 	SLV_ORDERS_OWNED_BY_ORDERLIST,          ///< 354  PR#13948 Orders stored in OrderList, pool removed.
 
+	SLV_INDUSTRY_ACCEPTED_HISTORY,          ///< 355  PR#14321 Industry acceptance history.
```
```suggestion
	SLV_INDUSTRY_ACCEPTED_HISTORY,          ///< 355  PR#14321 Industry history of cargo delivered and waiting.
```

Acceptance history sounds like an industry denying cargo deliveries, as I think ECS does. Better wording?

2TallTyler reviewed (commented):
Vanilla works great. I haven't tested with GRFs that stockpile yet, but your averaging sounds like it'll be fine.

rubidium42 commented on src/graph_gui.cpp:
```diff
@@ -216,6 +217,14 @@ struct BaseGraphWindow : Window {
+		DataSet &dataset; ///< Dataset to fill.
+		const Tprojection &proj; ///< Projection to apply.
+
+		inline void Fill(uint i, const auto &data) const { this->dataset.values[i] = std::invoke(this->proj, data); }
```
Why `uint` instead of `size_t`?

PeterN replied:
`size_t` felt unnecessary given the range. But then so is `uint`, I guess.

rubidium42 commented on src/misc/history_func.hpp:
```diff
@@ -0,0 +1,41 @@
+template <uint N, typename T, typename... Tfillers>
+void FillFromHistory(const HistoryData<T> &history, Tfillers... fillers)
+{
+	for (uint i = 0; i != N; ++i) {
```
Why `uint` instead of `size_t`?

PeterN committed:
Codechange: Allow unused graph ranges to be masked.

PeterN committed:
Add: Industry accepted and waiting history graphs.

Records amount of cargo accepted, and a rolling average of the waiting amount.

Average waiting samples the waiting amount once per day for each industry, spread out over an economy day.
